### PR TITLE
Hooks

### LINKED ISSUE #510: New Hooks React
React new hooks https://reactjs.org/docs/hooks-intro.html . Can we integrate them ScalaFnComponent
 
```
ScalaFnComponent[Props](props => {
```

### PR DESCRIPTION
**Pull request by japgolly**
Closes #510

### CODE REVIEW

File: core/src/main/scala/japgolly/scalajs/react/feature/Hooks.scala [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+  final case class UseState[S](raw: Raw.React.UseState[S]) {
+
+    @inline def state: S =
+      raw._1
+
+    def setState(s: S): Callback =
+      Callback(raw._2(s))
+
+    def modState(f: S => S): Callback =
+      Callback(modStateRaw(f))
```
Comment (rpiaggio) [inline core/src/main/scala/japgolly/scalajs/react/feature/Hooks.scala:174]: Thanks for the live session yesterday, it was really fun.

Thinking about ergonomics at the use site, maybe these 3 methods can be called `get`, `set` and `mod`.

File: core/src/main/scala/japgolly/scalajs/react/feature/Hooks.scala [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+      * escape hatch from React’s purely functional world into the imperative world.
+      *
+      * By default, effects run after every completed render.
+      * If you'd only like to execute the effect when your component is mounted, then use [[useEffectOnMount]].
+      * If you'd only like to execute the effect when certain values have changed, provide those certain values as
+      * the second argument.
+      *
+      * @see https://reactjs.org/docs/hooks-reference.html#useeffect
+      */
+    def useEffect[A](effect: CallbackTo[A])(implicit p: EffectArg[A]): Unit =
```
Comment (rpiaggio) [inline core/src/main/scala/japgolly/scalajs/react/feature/Hooks.scala:72]: The `effect` parameter could be a more Scala-like `Callback[Option[Callback]]`.

File: core/src/main/scala/japgolly/scalajs/react/feature/Hooks.scala (hunk 1 above)
Comment (japgolly) [reply core/src/main/scala/japgolly/scalajs/react/feature/Hooks.scala:251]: Yeah huge thanks for joining and for your input! It was fun! :grinning: 

Cool idea, let me think about it. I had a similar thought for existing API years ago and decided to stick with the `{state,setState,modState}` pattern.

File: core/src/main/scala/japgolly/scalajs/react/feature/Hooks.scala (hunk 2 above)
Comment (japgolly) [reply core/src/main/scala/japgolly/scalajs/react/feature/Hooks.scala:101]: Dude every time you open your mouth gold comes out! I need to run very soon but I should be able to do implement this quickly now... Clock-racing time!

Comment (japgolly) [reply core/src/main/scala/japgolly/scalajs/react/feature/Hooks.scala:107]: Done. See `optionalCallback`. Thanks!

File: core/src/main/scala/japgolly/scalajs/react/feature/Hooks.scala (hunk 1 above)
Comment (japgolly) [reply core/src/main/scala/japgolly/scalajs/react/feature/Hooks.scala]: Marking as resolved cos as we discussed on stream, I'll align this with the StateSnapshot api: `{value,setState,modState}`

File: core/src/main/scala/japgolly/scalajs/react/hooks/Api.scala [hunk 3]
```diff
@@ -0,0 +1,674 @@
...
+      * Prefer the standard [[useEffect]] when possible to avoid blocking visual updates.
+      *
+      * If you'd only like to execute the effect when your component is mounted, then use [[useLayoutEffectOnMount]].
+      * If you'd only like to execute the effect when certain values have changed, provide those certain values as
+      * the second argument.
+      *
+      * @see https://reactjs.org/docs/hooks-reference.html#useLayoutEffect
+      */
+    final def useLayoutEffectBy[A](init: CtxFn[CallbackTo[A]])(implicit a: UseEffectArg[A], step: Step): step.Self =
+      useEffectBy(step.squash(init)(_))
```
Comment (japgolly) [inline core/src/main/scala/japgolly/scalajs/react/hooks/Api.scala:594]: OI!

File: core/src/main/scala/japgolly/scalajs/react/hooks/Api.scala [hunk 4]
```diff
@@ -0,0 +1,674 @@
...
+      * Prefer the standard [[useEffect]] when possible to avoid blocking visual updates.
+      *
+      * If you'd only like to execute the effect when your component is mounted, then use [[useLayoutEffectOnMount]].
+      * If you'd only like to execute the effect when certain values have changed, provide those certain values as
+      * the second argument.
+      *
+      * @see https://reactjs.org/docs/hooks-reference.html#useLayoutEffect
+      */
+    final def useLayoutEffectBy[A](init: Ctx => CallbackTo[A])(implicit a: UseEffectArg[A], step: Step): step.Self =
+      self(ctx => UseEffect.unsafeCreate(init(ctx)))
```
Comment (japgolly) [inline core/src/main/scala/japgolly/scalajs/react/hooks/Api.scala:287]: OI!

File: core/src/main/scala/japgolly/scalajs/react/hooks/CustomHook.scala [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+  trait ArgLowPri {
+    implicit def id[A, B >: A]: Arg[A, B] = Arg(a => a)
+  }
+
+  object Arg extends ArgLowPri {
+    def const[C, I](i: I): Arg[C, I] =
+      apply(_ => i)
+
+    implicit def unit    [Ctx]      : Arg[Ctx, Unit]                       = const(())
+    implicit def ctxProps[P]        : Arg[HookCtx.P0[P], P]                = apply(_.props)
```
Comment (japgolly) [inline core/src/main/scala/japgolly/scalajs/react/hooks/CustomHook.scala:59]: test this

File: core/src/main/scala/japgolly/scalajs/react/hooks/HookComponentBuilder.scala [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+
+    final class First[P](init: P => Unit) extends Api.Primary[P, FirstStep[P]] {
+
+      override protected def self(f: P => Any)(implicit step: Step): step.Self =
+        step.self(init, f)
+
+      override protected def next[H](f: P => H)(implicit step: Step): step.Next[H] =
+        step.next(init, f)
+
+      def withPropsChildren: ComponentPC.First[P] =
```
Comment (japgolly) [inline core/src/main/scala/japgolly/scalajs/react/hooks/HookComponentBuilder.scala:27]: Document :(

File: core/src/main/scala/japgolly/scalajs/react/hooks/UseCallbackBoilerplate.scala [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+    UseCallbackArg[(A, B, C, D, E, F, G, H, I, J, K, L, M, N) => Callback, js.Function14[A, B, C, D, E, F, G, H, I, J, K, L, M, N, Unit]](
+      f => f(_, _, _, _, _, _, _, _, _, _, _, _, _, _).runNow())(
+      z => Reusable.byRef(z).withValue((a, b, c, d, e, f, g, h, i, j, k, l, m, n) => Callback(z(a, b, c, d, e, f, g, h, i, j, k, l, m, n))))
+
+
+  implicit def c15[A, B, C, D, E, F, G, H, I, J, K, L, M, N, O]: UseCallbackArg[(A, B, C, D, E, F, G, H, I, J, K, L, M, N, O) => Callback] =
+    UseCallbackArg[(A, B, C, D, E, F, G, H, I, J, K, L, M, N, O) => Callback, js.Function15[A, B, C, D, E, F, G, H, I, J, K, L, M, N, O, Unit]](
+      f => f(_, _, _, _, _, _, _, _, _, _, _, _, _, _, _).runNow())(
+      z => Reusable.byRef(z).withValue((a, b, c, d, e, f, g, h, i, j, k, l, m, n, o) => Callback(z(a, b, c, d, e, f, g, h, i, j, k, l, m, n, o))))
+
```
Comment (japgolly) [inline core/src/main/scala/japgolly/scalajs/react/hooks/UseCallbackBoilerplate.scala:105]: double blank lines

File: doc/HOOKS.md [hunk 8]
```diff
@@ -0,0 +1,542 @@
...
+  val hook2: CustomHook[Unit, Int] = ???
+  val hooks: CustomHook[Unit, (Int, Int)] = hook1 ++ hook2
+}
+
+object Example3 {
+  val hook1: CustomHook[Long, Boolean] = ???
+  val hook2: CustomHook[String, Int] = ???
+  val hooks: CustomHook[(Long, String), (Boolean, Int)] = hook1 ++ hook2
+}
+```
```
Comment (japgolly) [inline doc/HOOKS.md:291]: Add another example with a Unit on both sides (i.e. AI, BO)

**Pull request merged**